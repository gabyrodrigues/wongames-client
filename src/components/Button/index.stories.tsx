import { StoryObj, Meta } from "@storybook/react";
import Button, { ButtonProps } from ".";

export default {
  title: "Button",
  component: Button,
  argTypes: {
    children: {
      type: "string"
    }
  }
} as Meta<ButtonProps>;

export const Default: StoryObj = {
  args: {
    children: "Buy now"
  }
};
