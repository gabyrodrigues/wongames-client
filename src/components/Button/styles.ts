import styled, { css } from "styled-components";
import { ButtonProps } from ".";
import { Theme } from "@styles/theme";
import { darken } from "polished";

interface WrapperProps extends Omit<ButtonProps, "children"> {
  $fullWidth?: boolean;
  $hasIcon: boolean;
  $minimal?: boolean;
}

const wrapperModifiers = {
  small: (theme: Theme) => css`
    font-size: ${theme.font.sizes.xsmall};
    height: 3rem;
  `,
  medium: (theme: Theme) => css`
    font-size: ${theme.font.sizes.small};
    height: 4rem;
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,
  large: (theme: Theme) => css`
    font-size: ${theme.font.sizes.medium};
    height: 5rem;
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
  `,
  fullWidth: () => css`
    width: 100%;
  `,
  withIcon: (theme: Theme) => css`
    svg {
      width: 1.5rem;

      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `,
  minimal: (theme: Theme) => css`
    background: none;
    color: ${theme.colors.primary};

    &:hover {
      color: ${darken(0.1, theme.colors.primary)};
    }
  `
};

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, $fullWidth, $hasIcon, $minimal }) => css`
    align-items: center;
    background: linear-gradient(180deg, #ff5f5f 0%, #f062c0 50%);
    border: 0;
    border-radius: ${theme.border.radius};
    color: ${theme.colors.white};
    cursor: pointer;
    display: inline-flex;
    justify-content: center;
    padding: ${theme.spacings.xxsmall};
    text-decoration: none;

    &:hover {
      background: ${$minimal ? "none" : "linear-gradient(180deg, #e35565 0%, #d958a6 50%)"};
    }

    ${!!size && wrapperModifiers[size](theme)}
    ${!!$fullWidth && wrapperModifiers.fullWidth}
    ${!!$hasIcon && wrapperModifiers.withIcon(theme)}
    ${!!$minimal && wrapperModifiers.minimal(theme)}
  `}
`;
