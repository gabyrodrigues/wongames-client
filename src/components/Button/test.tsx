import { screen } from "@testing-library/react";
import Button from ".";
import { renderWithTheme } from "@utils/tests/helpers";
import { AddShoppingCart } from "@styled-icons/material-outlined";

describe("<Button />", () => {
  it("should render the medium size by default", () => {
    const { container } = renderWithTheme(<Button>Buy now</Button>);

    expect(screen.getByRole("button", { name: /Buy now/i })).toHaveStyle({
      fontSize: "1.4rem",
      height: "4rem",
      padding: "0.8rem 3.2rem"
    });

    expect(container.firstChild).toMatchSnapshot();
  });
  it("should render the small size", () => {
    renderWithTheme(<Button size="small">Buy now</Button>);

    expect(screen.getByRole("button", { name: /Buy now/i })).toHaveStyle({
      fontSize: "1.2rem",
      height: "3rem"
    });
  });
  it("should render the large size", () => {
    renderWithTheme(<Button size="large">Buy now</Button>);

    expect(screen.getByRole("button", { name: /Buy now/i })).toHaveStyle({
      fontSize: "1.6rem",
      height: "5rem",
      padding: "0.8rem 4.8rem"
    });
  });
  it("should render a fullWidth version", () => {
    renderWithTheme(<Button fullWidth>Buy now</Button>);

    expect(screen.getByRole("button", { name: /Buy now/i })).toHaveStyle({
      width: "100%"
    });
  });
  it("should render an icon version", () => {
    renderWithTheme(<Button icon={<AddShoppingCart data-testid="icon" />}>Buy now</Button>);

    expect(screen.getByText(/Buy now/i)).toBeInTheDocument();
    expect(screen.getByTestId("icon")).toBeInTheDocument();
  });
  it("should render a minimal version", () => {
    renderWithTheme(
      <Button
        icon={<AddShoppingCart data-testid="icon" />}
        minimal>
        Buy now
      </Button>
    );
    expect(screen.getByRole("button", { name: /Buy now/i })).toHaveStyle({
      background: "none",
      color: "#F231A5"
    });
    expect(screen.getByRole("button", { name: /buy now/i })).toHaveStyleRule("background", "none", {
      modifier: "hover"
    });
  });
  it("should render Button as a link", () => {
    renderWithTheme(
      <Button
        as="a"
        href="/link">
        Buy now
      </Button>
    );

    expect(screen.getByRole("link", { name: /buy now/i })).toHaveAttribute("href", "/link");
  });
});
