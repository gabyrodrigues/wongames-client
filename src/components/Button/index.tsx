import { ReactNode } from "react";
import * as S from "./styles";

export type ButtonProps = {
  children?: ReactNode;
  size?: "small" | "medium" | "large";
  fullWidth?: boolean;
  icon?: ReactNode;
  onClick?: () => (event: React.MouseEvent<HTMLButtonElement>) => void;
};

export default function Button({
  children,
  size = "medium",
  fullWidth = false,
  icon,
  ...props
}: ButtonProps) {
  return (
    <S.Wrapper
      size={size}
      $fullWidth={fullWidth}
      $hasIcon={!!icon}
      {...props}>
      {!!icon && icon}
      {!!children && <span>{children}</span>}
    </S.Wrapper>
  );
}
