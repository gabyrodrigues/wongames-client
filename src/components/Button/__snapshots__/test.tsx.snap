// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render a minimal version 1`] = `
.c0 {
  align-items: center;
  background: linear-gradient(180deg, #ff5f5f 0%, #f062c0 50%);
  border: 0;
  border-radius: 0.4rem;
  color: #FAFAFA;
  cursor: pointer;
  display: inline-flex;
  justify-content: center;
  padding: 0.8rem;
  text-decoration: none;
  font-size: 1.4rem;
  height: 4rem;
  padding: 0.8rem 3.2rem;
  background: none;
  color: #F231A5;
}

.c0 a {
  text-decoration: none;
  color: #FAFAFA;
}

.c0:hover {
  background: none;
}

.c0 svg {
  width: 1.5rem;
}

.c0 svg+span {
  margin-left: 0.8rem;
}

.c0:hover {
  color: #e20e8d;
}

.c1 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

<button
  class="c0"
>
  <svg
    aria-hidden="true"
    class="c1"
    data-testid="icon"
    fill="currentColor"
    focusable="false"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-8.9-5h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4l-3.87 7H8.53L4.27 2H1v2h2l3.6 7.59-1.35 2.44C4.52 15.37 5.48 17 7 17h12v-2H7l1.1-2z"
    />
  </svg>
  <span>
    Buy now
  </span>
</button>
`;

exports[`<Button /> should render the medium size by default 1`] = `
.c0 {
  align-items: center;
  background: linear-gradient(180deg, #ff5f5f 0%, #f062c0 50%);
  border: 0;
  border-radius: 0.4rem;
  color: #FAFAFA;
  cursor: pointer;
  display: inline-flex;
  justify-content: center;
  padding: 0.8rem;
  text-decoration: none;
  font-size: 1.4rem;
  height: 4rem;
  padding: 0.8rem 3.2rem;
}

.c0 a {
  text-decoration: none;
  color: #FAFAFA;
}

.c0:hover {
  background: linear-gradient(180deg, #e35565 0%, #d958a6 50%);
}

<button
  class="c0"
>
  <span>
    Buy now
  </span>
</button>
`;
